content/en/docs/setup/debug.md

```diff
@@ -177,14 +173,139 @@ Events:
+Users may encounter the below errors when running `keadm join` command:
+```
+edge node join failed: pull Images failed: xxx
+```
+Troubleshooting ideas:
+
+- Can edge nodes access the Internet? `keadm join` will pull `kubeedge/installation-package` image, from dockerhub by default. If so, users should fix according to the specific error information.
+- If edge nodes cannot access the Internet, there're two ways to avoid this error:
+    1. download images on another machines which can access the Internet, and load images to the edge nodes
+    2. download images on another machines which can access the Internet, and push images to the internal image repository, and then run `keadm join` command with flag `--image-repository=xxx` to specify the image repository.
+
+
+## Token related issues
+
+### Error: token credentials are in the wrong format
+
+If you find that edgecore don't start, and the edgecore log contains error message like below
+```shell
+F1121 14:14:39.583329 3644556 certmanager.go:94] Error: token credentials are in the wrong format
+```
+This means the edgecore token is not correct. You should run `keadm gettoken` in cloud side to get token, and then copy it to `modules.edgeHub.token` field.
+```shell
+# cat /etc/kubeedge/config/edgecore.yaml 
+apiVersion: edgecore.config.kubeedge.io/v1alpha2
+kind: EdgeCore
+modules:
+  edgeHub:
+    ...
+    ...
+    token: ""      # --- here is token, which will be used when joining edge nodes to cloud.
+
+```
+
+### Invalid Token
+If edgecore log contains errors like below
+```shell
+F1121 14:22:42.419103 3646881 certmanager.go:94] Error: failed to get edge certificate from the cloudcore, error: Invalid authorization token
+```
+And cloudcore log contains errors like below
+```shell
+E1121 06:22:42.418947       1 server.go:104] failed to sign the certificate for edgenode: edge-node, invalid token
+```
+This means that you just copy a NOT correct token to edgecore configuration. So you should run `keadm gettoken` in cloud side to get token again, and copy it to `modules.edgeHub.token` field rightly and carefully.
+
+
+## edgecore failed to pass the certificate verification
+
+```
+Error: failed to get edge certificate from the cloudcore, error: Get "https://192.168.47.128:10002/edge.crt": x509: certificate is valid for 192.168.47.127, not 192.168.47.128
+```
+
+```
+Error: failed to get edge certificate from the cloudcore, error: Get "https://192.168.47.128:10002/edge.crt": x509: cannot validate certificate for 192.168.47.128 because it doesn't contain any IP SANs
+```
+
+cloudcore need configuring advertise-address, which can contain multiple IP addresses, splitted with commas, which need to be planned in advance. It's recommended to use load balancer or gateway address, to ensure load balancing and high availability of cloudcore. Concurrently, this address cannot be changed once it's configured. The value of `cloudcore-ipport` of edgecore must exist in the cloudcore `advertise-address` list.
```

Comment:
rephrasing

Reply:
```suggestion
For more details about taints and tolerations, you can refer to [k8s official docs](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/)


### cloudcore in ImagePullBackOff status
If the pod `cloudcore` is in `ImagePullBackOff` status, it's often due to pod image issues. You can run `kubectl describe` command to find more details. For example, the output may be showing like below.

```shell
# kubectl describe pod cloudcore-644d8f55df-sj7xc -nkubeedge

...
...

Events:
  Type     Reason     Age                     From               Message
  ----     ------     ----                    ----               -------
  Normal   Scheduled  8m39s                   default-scheduler  Successfully assigned kubeedge/cloudcore-644d8f55df-sj7xc to kind-control-plane
  Normal   Pulling    6m59s (x4 over 8m39s)   kubelet            Pulling image "kubeedge/cloudcore:v1.12.19"
  Warning  Failed     6m58s (x4 over 8m37s)   kubelet            Failed to pull image "kubeedge/cloudcore:v1.12.19": rpc error: code = NotFound desc = failed to pull and unpack image "docker.io/kubeedge/cloudcore:v1.12.19": failed to resolve reference "docker.io/kubeedge/cloudcore:v1.12.19": docker.io/kubeedge/cloudcore:v1.12.19: not found
  Warning  Failed     6m58s (x4 over 8m37s)   kubelet            Error: ErrImagePull
  Warning  Failed     6m45s (x6 over 8m37s)   kubelet            Error: ImagePullBackOff
  Normal   BackOff    3m36s (x20 over 8m37s)  kubelet            Back-off pulling image "kubeedge/cloudcore:v1.12.19"
```
Here we can find the core reason why cloudcore did't start successfully. It's due to `Failed to pull image "kubeedge/cloudcore:v1.12.19": rpc error: code = NotFound desc = failed to pull and unpack image "docker.io/kubeedge/cloudcore:v1.12.19": failed to resolve reference "docker.io/kubeedge/cloudcore:v1.12.19": docker.io/kubeedge/cloudcore:v1.12.19: not found`. There's no `docker.io/kubeedge/cloudcore:v1.12.19` image. In other words, we specified a wrong kubeedge version, which does not exist. So we can run `keadm reset` to clear the last installation, or just run `kubectl delete ns kubeedge` directly, and then run `keadm init` to start a new installation, with a correct version.

During the cloudcore container installation process, keadm will pull the required images from dockerhub by default. Users can pull the required images in advance, which is also supported by `keadm`. To do so, users can run `keadm config images` related commands to display or pull all the required images.
```shell
# keadm config images pull --kubeedge-version=v1.12.0 --part=cloud
Pulling kubeedge/iptables-manager:v1.12.0 ...
Successfully pulled kubeedge/iptables-manager:v1.12.0
Pulling kubeedge/controller-manager:v1.12.0 ...
Successfully pulled kubeedge/controller-manager:v1.12.0
Pulling kubeedge/admission:v1.12.0 ...
Successfully pulled kubeedge/admission:v1.12.0
Pulling kubeedge/cloudcore:v1.12.0 ...
Successfully pulled kubeedge/cloudcore:v1.12.0
```

### cloudcore in CrashLoopBackOff status
If `cloudcore` pod is in `CrashLoopBackOff` status, it means that cloudcore pod startup failed. Users should check cloudcore pod logs by running `kubectl logs` command. And fix it according to the log error.

## EdgeCore failed to get CA certificate
During the edgecore installation, edgecore will communicate with cloudcore to get the CA certificate. Users can encounter the below issues:
```shell
failed to get CA certificate, err: Get "https://192.168.47.128:10002/ca.crt": dial tcp 192.168.47.128:10002: connect: connection refused
```

```shell
failed to get CA certificate, err: Get "https://192.168.47.128:10002/ca.crt": EOF
```

```shell
Error: failed to get CA certificate, err: Get "https://192.168.47.128:10002/ca.crt": dial tcp 192.168.47.128:10002: connect: no route to host
```

```shell
failed to get CA certificate, err: Get "https://192.168.47.128:10002/ca.crt": dial tcp 192.168.47.128:10002: i/o timeout
```

Troubleshooting steps:

- Does the cloudcore IP address, the edge node is configured to join to, exist in the cloudcore advertise-address list?

- Can the edgecore connect to the cloudcore IP address? Are the two nodes connected over the physical network? Are there firewall restrictions?

- Do cloudcore components on the cloud side start successfully and are the accessed ports (e.g. 10000 and 10002) listening?

- If all of the above preceding steps were checked and are not showing any problems, run the following command to check whether cloudcore logs contain any errors: `kubectl logs cloudcore-xxxx -n kubeedge`

## `keadm join` pull image failed in edge nodes
Users may encounter the below errors when running `keadm join` command:
```
edge node join failed: pull Images failed: xxx
```
Troubleshooting steps:

- Can the edge nodes access the Internet? `keadm join` will by default pull `kubeedge/installation-package` image, from dockerhub. If so, users should fix according to the specific error information.
- If edge nodes cannot access the Internet, there are two ways to avoid this error:
    1. download images on another machine which can access the Internet, and load images manually to the edge nodes
    2. download images on another machine which can access the Internet, and push images to an internal image repository, and then run `keadm join` command with flag `--image-repository=xxx` to specify the internal image repository.


## Token related issues

### Error: token credentials are in the wrong format

If you find that edgecore don't start, and the edgecore log contains error message like below
```shell
F1121 14:14:39.583329 3644556 certmanager.go:94] Error: token credentials are in the wrong format
```
This means the edgecore token is not correct. You should run `keadm gettoken` on the cloud side to get token, and then copy it to `modules.edgeHub.token` field.
```shell
# cat /etc/kubeedge/config/edgecore.yaml 
apiVersion: edgecore.config.kubeedge.io/v1alpha2
kind: EdgeCore
modules:
  edgeHub:
    ...
    ...
    token: ""      # --- here is token, which will be used when joining edge nodes to cloud.

```

### Invalid Token
If edgecore log contains errors like below
```shell
F1121 14:22:42.419103 3646881 certmanager.go:94] Error: failed to get edge certificate from the cloudcore, error: Invalid authorization token
```
And cloudcore log contains errors like below
```shell
E1121 06:22:42.418947       1 server.go:104] failed to sign the certificate for edgenode: edge-node, invalid token
```
This means that you copied an INCORRECT token to the edgecore configuration. So you should run `keadm gettoken` on the cloud side to get the token again, and copy it to `modules.edgeHub.token` field correctly and carefully.


## edgecore failed to pass the certificate verification

```
Error: failed to get edge certificate from the cloudcore, error: Get "https://192.168.47.128:10002/edge.crt": x509: certificate is valid for 192.168.47.127, not 192.168.47.128
```

```
Error: failed to get edge certificate from the cloudcore, error: Get "https://192.168.47.128:10002/edge.crt": x509: cannot validate certificate for 192.168.47.128 because it doesn't contain any IP SANs
```

Cloudcore requires the configuration parameter advertise-address, which can contain multiple IP addresses, separated by commas, which need to be defined in advance. It's recommended to use a load balancer or gateway address, to enable load balancing and high availability of cloudcore. Currently, this address cannot be changed once it's configured. The value of `cloudcore-ipport` at edgecore must exist in the cloudcore `advertise-address` list.
```